### PR TITLE
Vendor JS moved into webpack

### PR DESCRIPTION
**Pull request by jdlrobson**
The script static/js/vendor.jsh previously constructed a vendor.js
bundle that was shipped to users. This meant every page had 2 JS
files to download. This moves all those libraries verbatim, using
npm pinned versions where possible to webpack and imports them all
in the exact same order at the top of index.js

This has 2 benefits:
* We now only load 1 JS file

This also enables me to do 3 follow ups:
* identify jquery migration issues in the console and fix them
* move vendor bundles closer to where they are actually used
* investigate upgrading libraries
* lazy load code on pages where it's not needed

### Description
This is a non-functional change. The way we generate code has changed, not the code itself.

Closes #

### Technical
Hopefully patch is self explanatory, 
The contents of static/js/vendor.jsh have moved to openlibrary/plugins/openlibrary/js/index.js - now webpack generates all our JS.
Several libraries are now loaded from npm pinned versions.

### Testing
- you'll need to run npm install and `make js` to see this in action.
- confirm only one JS file is now loaded per page
- I've checked various pages - subject, homepage, books page, edits page for JS errors and haven't seen any. Note you may notice some warnings in the console from jquery migrate - it's great to finally see these so we can give them our attention! Another pair of eyes doing the same would be useful.

### Evidence
There should be no UI changes with this change.

### CODE REVIEW

**Comment by cdrini** [comment]
Did some perf tests; time is computed using Firefox's throttling for "Regular 3G", averaged over 5 runs, with caching disabled. Tested on page http://192.168.99.100:8080/books/OL2058361M/The_wit_wisdom_of_Mark_Twain

Before: 

| File         | Transferred | Size    | Time  |
| ------------ | --------- | --------- | ----- |
| vendor-v2.js | 575.76 KB | 575.54 KB | 6.7 s |
| all.js       | 133.10 KB | 132.87 KB | 5.0 s |
| Net          | 708.86 KB | 708.39 KB | 6.7 s |

After:
| File         | Transferred | Size    | Time  |
| ------------ | --------- | --------- | ----- |
| all.js       | 669.04 KB | 669.04 KB | 7.1 s |
| Net          | 669.04 KB | 669.04 KB | 7.1 s |

We don't have gzipping enabled locally (unfortunately), so the real time will be less though, so I don't think it will make much of a difference, so perf is ok.

**Comment by cdrini** [comment]
It looks like the markdown previews are broken now; maybe wmd needs to be loaded in index.js?
image

**Comment by cdrini** [comment]
image

😍

File: webpack.config.js [hunk 1]
```diff
@@ -78,6 +84,8 @@ module.exports = {
     devtool: 'source-map',
 
     performance: {
+        maxAssetSize: 670 * 1024,
+        maxEntrypointSize: 670 * 1024,
```
Comment (cdrini) [inline webpack.config.js]: Can you add comments about what these lines do?

Comment (jdlrobson) [reply webpack.config.js]: What would it say? The whole config is documented on https://webpack.js.org/configuration/performance/

The default value is 250kb which we are now exceeding, so we have to explicitly define these.

**Comment by jdlrobson** [comment]
> It looks like the markdown previews are broken now; maybe wmd needs to be loaded in index.js?
> image
> ^ http://localhost:8080/works/OL53924W/The_complete_works_of_Mark_Twain/edit

Yeh looks like I didn't copy that over. Will do.

Regarding gzip, bundlesize will tell you this - and we define in package.json . Before vendor-v2.js was 164KB with gzipping and all.js was 35.5KB. The combined gzip now is 201.2kb (I'm a bit surprised as I could have sworn it went down when first worked on this patch but it's only a minor increase, and having one bundle will allow us to code split which on the long term will speed things up considerably - see a2f27fec58d74145677547a4674d5f6e361be1fe and eb97d05b245cd175ebdb561399359d229562cf56 for a sneak peak)

File: webpack.config.js (hunk 1 above)
Comment (cdrini) [reply webpack.config.js]: That's fair; maybe add that link to the `performance` object? Build config files are kind of difficult to grok, so making it easier for future folk to understand what's happening makes them less intimidating.

File: openlibrary/plugins/openlibrary/js/index.js [hunk 2]
```diff
@@ -1,3 +1,28 @@
...
+// jquery-show-password#1.0 not on npm, no longer getting worked on
+import '../../../../vendor/js/jquery-showpassword/jquery.showpassword.js';
+// jquery.form#2.36 not on npm, no longer getting worked on
+import '../../../../vendor/js/jquery-form/jquery.form.js';
+// jquery-validate#1.6 not on npm
+import '../../../../vendor/js/jquery-validate/jquery.validate.js';
+// jquery-autocomplete#1.1 with modified
+import '../../../../vendor/js/jquery-autocomplete/jquery.autocomplete-modified.js';
+// jquery-flot 0.7.0
+import '../../../../vendor/js/wmd/jquery.wmd.min.js'
```
Comment (cdrini) [inline openlibrary/plugins/openlibrary/js/index.js]: This fell into the flot section; Can you add a comment like you've done for the others?

**Comment by jdlrobson** [comment]
Would it be better to link to the entire docs at the top of the file? I'm
just a bit worried we will have lots of lines pointing to sub pages that
might change URL with future webpack releases.

Agree that config is difficult to grok. We have a dev at wikimedia we call
the webpack guy 😂

File: webpack.config.js (hunk 1 above)
Comment (jdlrobson) [reply webpack.config.js]: I've linked to https://webpack.js.org/configuration. See comment below

**Comment by jdlrobson** [comment]
> The markdown previews are still not working; are they working for you?
> 
> (Also, those sneak peak commits look glorious :) )

Okay so I debugged this. The jquery.wmd library does this:
```
var Attacklab = Attacklab || {}
```
it expects that to register a global, which is no longer the case, which breaks everything.
To fix this we'll need to add:
```
window.Attacklab = Attacklab
```
at the bottom of jquery.wmd

This can be done with https://github.com/internetarchive/wmd/pull/4 and an update to vendor in the codebase. Will need help from @mekarpeles on this one. Then this patch should just work.

**Comment by cdrini** [comment]
In the interest of getting this out the door, could we include the `wmd` file in a `<script>` tag directly and remove the JS import? I want to merge this PR :)

(I'm also nervous to edit the wmd code and would rather avoid it if possible :P)

**Comment by jdlrobson** [comment]
> In the interest of getting this out the door, could we include the `wmd` file in a `<script>` tag directly and remove the JS import? I want to merge this PR and take it off my to-do list :)
> 
> (I'm also nervous to edit the wmd code and would rather avoid it if possible :P)

vendor is not publicly accessible, so to do this we'd need to keep that Vendor make file to copy it across. We'd then have to load it on every page with an additional request (after the core JS file) I'd rather not do that as the goal of this is to load less on every page.

The quickest short-term fix I can imagine that I'm okay with  is doing this:
```
cp vendor/js/wmd/jquery.wmd.js openlibrary/plugins/openlibrary/js/
```

After that
1) update the import in index.js to this file
2) add
```
window.Attacklab = Attacklab;
```
to the bottom of the copied file.

Thoughts?

**Comment by jdlrobson** [comment]
Okay this should be working now. Beware Docker filesystem issues (I'm hitting some myself). You may need to cd vendor/js/wmd and do git fetch origin && git checkout origin/master in docker bash if wmd is not working.

**Comment by jdlrobson** [comment]
Screenshot 2019-08-19 at 5 57 43 PM
Proof it's working ^

**Comment by cdrini** [comment]
Yeah, that is weird! I think it's because we have a volume for `vendor/`. Looks like if you run `docker-compose exec web git submodule update --recursive` will work, but it won't work if you updated the submodules outside of docker.

**Pull request merged**